File: 프로그래머스/lv4/42897. 도둑질/도둑질.cpp
#include <string>
#include <vector>
#include <algorithm>

using namespace std;

int solution(vector<int> money) {
    int answer = 0;
    int len = money.size();
    
    if(len == 1) {
        answer = money[0];
    }
    else if(len == 2) {
        answer = max(money[0], money[1]);
    }
    else{
        vector<int> dp1(len, 0);
        vector<int> dp2(len, 0);
        dp1[0] = money[0];
        // 0번째 집 무조건 포함.
        for(int i = 1; i < len - 1; i++) {
            dp1[i] = max(dp1[i - 1], dp1[i - 2] + money[i]);
        }
        // 0번째 집 무조건 포함 안함.
        for(int i = 1; i < len; i++) {
            dp2[i] = max(dp2[i - 1], dp2[i - 2] + money[i]);
        }

        answer = max(dp1[len - 2], dp2[len - 1]);   
    }
    
    return answer;
}
// 어렵넹..

File: 프로그래머스/4/42897. 도둑질/도둑질.py
def solution(money):
    answer = 0
    
    N = len(money)
    
    if N == 1 or N == 2:
        return max(money)
    
    dp = [[0] * N for _ in range(2)]
    
    dp[0][0] = money[0]
    for idx in range(1, N - 1):
        dp[0][idx] = max(dp[0][idx - 1], dp[0][idx - 2] + money[idx])
    for idx in range(1, N):
        dp[1][idx] = max(dp[1][idx - 1], dp[1][idx - 2] + money[idx])
        
    return max(dp[0][N - 2], dp[1][N - 1])